### PR TITLE
[Feature] Add ability to create schedules from existing transactions

### PR DESCRIPTION
**Pull request by xentara1**
#2221 
Please let me know thoughts, or if you need any additional information

### CODE REVIEW

**Comment by MatissJanis** [comment]
👋 Cool feature!

The list of all transaction options is getting quite long. Could you instead create a "create schedule" button in the "Link schedule" modal? That would solve the same problem (albeit with one more click, but I think that's fine since we don't expect people to be creating schedules often).

**Comment by xentara1** [comment]
Adjusted let me know thoughts.

**Comment by youngcw** [comment]
Nice.  I like that new placement of the button.  We probably should add this to the docs in some way

**Comment by xentara1** [comment]
Fixed lint's, 
Not used typescript before so don't think its the best way hah.

**Comment by xentara1** [comment]
> Nice. I like that new placement of the button. We probably should add this to the docs in some way

Where do I update that?

**Comment by youngcw** [comment]
> Fixed lint's, Not used typescript before so don't think its the best way hah.

if you run `yarn lint --fix` most lint issues will get taken care of

**Comment by youngcw** [comment]
> Where do I update that?

There is a repo for the docs and website

**Comment by xentara1** [comment]
> > Where do I update that?
> 
> There is a repo for the docs and website https://github.com/actualbudget/docs

Added -> 
https://github.com/actualbudget/docs/pull/303

**Comment by MatissJanis** [comment]
Nice! This looks much better. But I'm seeing some usability problems:

- [ ] after creating the schedule - it should auto-update the selected transaction so it would be added to the new schedule
- [ ] after creating the schedule - all modals should be closed

Would you mind taking a look at these small issues? Also: please run `yarn lint --fix`

Let me know if I/we can help!

**Comment by xentara1** [comment]
Makes sense, adjusted as per comments, 
Removed a check on event as don't see why it is a must, let me know if I'm missing something

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx [hunk 1]
```diff
@@ -44,6 +55,14 @@ export function ScheduleLink({
     }
     actions.popModal();
   }
+  async function onCreate() {
+    const firstId1 = [...ids][0];
+    actions.popModal();
+    pushModal('schedule-edit', {
+      id: null,
+      transaction: getTransaction(firstId1),
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleLink.tsx:63]: **💬 suggestion:** ‏not quite sure why you would need to spread the `ids` to access 0-index.

```suggestion
    actions.popModal();
    pushModal('schedule-edit', {
      id: null,
      transaction: getTransaction(ids[0]),
```

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx [hunk 2]
```diff
@@ -4,24 +4,35 @@ import React, { useCallback, useRef, useState } from 'react';
...
   modalProps,
   actions,
   transactionIds: ids,
+  getTransaction,
+  pushModal,
 }: {
   actions: BoundActions;
   modalProps?: CommonModalProps;
   transactionIds: string[];
+  getTransaction: (a: string) => TransactionEntity;
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleLink.tsx:34]: **💬 suggestion:** ‏what's `a`? We should be more explicit
```suggestion
  getTransaction: (id: string) => TransactionEntity;
```

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx [hunk 3]
```diff
@@ -4,24 +4,35 @@ import React, { useCallback, useRef, useState } from 'react';
...
+  pushModal: (a: string, b: ModalParams) => void;
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleLink.tsx:35]: **💬 suggestion:** ‏same issue here: what's `a` and `b`?

File: packages/desktop-client/src/components/transactions/SelectedTransactions.jsx [hunk 4]
```diff
@@ -135,6 +135,8 @@ export function SelectedTransactionsButton({
           case 'link-schedule':
             pushModal('schedule-link', {
               transactionIds: [...selectedItems],
+              getTransaction : getTransaction,
+              pushModal: pushModal,
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/transactions/SelectedTransactions.jsx:139]: **🥜 nitpick:** ‏I'm surprised linter didn't pick this up
```suggestion
              getTransaction,
              pushModal,
```

File: packages/desktop-client/src/hooks/useSelected.tsx [hunk 5]
```diff
@@ -274,9 +273,6 @@ export function SelectedProvider<T extends Item>({
 
   const dispatch = useCallback(
     async (action: Actions) => {
-      if (!action.event && isNonProductionEnvironment()) {
-        throw new Error('SelectedDispatch actions must have an event');
-      }
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useSelected.tsx:279]: **❓ question:** ‏why are you changing this? Seems totally unrelated to the objective of this PR.

File: packages/desktop-client/src/hooks/useSelected.tsx [hunk 6]
```diff
@@ -65,7 +64,7 @@ export function useSelected<T extends Item>(
           const selectedItems = new Set(state.selectedItems);
           const { id, event } = action;
 
-          if (event.shiftKey && selectedRange) {
+          if (event && event.shiftKey && selectedRange) {
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useSelected.tsx:67]: **❓ question:** ‏why are you changing this? Is it causing any issues? If so: how can I reproduce?

File: packages/desktop-client/src/components/transactions/SimpleTransactionsTable.jsx [hunk 7]
```diff
@@ -155,7 +153,12 @@ export function SimpleTransactionsTable({
   const memoFields = useMemo(() => fields, [JSON.stringify(fields)]);
 
   const serializedTransactions = useMemo(() => {
-    return transactions.map(trans => serializeTransaction(trans, dateFormat));
+    transactions = transactions.map(trans =>
+      serializeTransaction(trans, dateFormat),
+    );
+    const selected = transactions.filter(x => x.selected).map(x => x.id);
+    selected.forEach(x => dispatchSelected({ type: 'select', id: x }));
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/transactions/SimpleTransactionsTable.jsx:160]: **🔨 warning:** ‏actions (`dispatch`) should not be called in memoization. Instead: use `useEffect` if you need to react to the data changing.

**Review by MatissJanis — CHANGES_REQUESTED** [review]
UX looks and works great! Did a first pass code review. Some smaller issues, but some bigger. Overall looks pretty good. Thanks!

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx (hunk 2 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleLink.tsx:34]: Fair cop, Was being lazy

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx (hunk 1 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleLink.tsx:63]: True, removed

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx (hunk 3 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleLink.tsx:35]: Same as above, updated

File: packages/desktop-client/src/components/transactions/SelectedTransactions.jsx (hunk 4 above)
Comment (xentara1) [reply packages/desktop-client/src/components/transactions/SelectedTransactions.jsx:139]: Adjusted as per suggestions, 
Is object notation not Key Value, 
I am not normally a JS/TS developer so to me this looks weird.

File: packages/desktop-client/src/hooks/useSelected.tsx (hunk 5 above)
Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: To Select an item in the grid(i.e. have the found transaction pre selected) I need to dispatch the "Select" Action
This has historically only been done from an Event action, i.e. mouse click in the grid. 

As i need to call it from a useEffect(prior memoziation) this is called without an Event,
Hence this bit of code needs to be removed.

File: packages/desktop-client/src/hooks/useSelected.tsx (hunk 6 above)
Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx:67]: As the if !action.event check is removed above, event can now be null, and thus this section now requires a null check. 
(lines:277-279)

File: packages/desktop-client/src/components/transactions/SimpleTransactionsTable.jsx (hunk 7 above)
Comment (xentara1) [reply packages/desktop-client/src/components/transactions/SimpleTransactionsTable.jsx]: Adjusted, 
&& Learned what memoization is...

**Comment by xentara1** [comment]
Updated all comments, 
Let me know if you want me to resolve the conversation, not sure how it normally works for opensource, 
When code reviewing at work, it annoys me when someone else resolves everything without me confirming,

File: packages/desktop-client/src/hooks/useSelected.tsx (hunk 5 above)
Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: Hmm.. do we need to update the selected items when creating a new schedule?

I'm really worried about what unintended consequences this particular change might have in other parts of the app, so ideally I would not like to see it changed if possible. If that means: the selected items get un-checked after creating a schedule - I think that's a good compromise.

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: **Reason I don't want to change:**
If we don't change it then workflow wise its weird, we are either create a new schedule and adding a transaction to it without visualising that in the UI.
||
We are creating a schedule from a transaction, and then not assigning that transaction to the schedule by default. 

Of the two above if we 100% dont want to change this code i would go with optiion 2, 

**Reason I think we don't need to be worried about this change:**
<details>
  <summary>Current Check Occurs Here</summary>
`  const dispatch = useCallback(
    async (action: Actions) => {
    if (!action.event && isNonProductionEnvironment()) {
        throw new Error('SelectedDispatch actions must have an event');
      }
      if (action.type === 'select-all') {
        if (!action.event && isNonProductionEnvironment()) {
          throw new Error('SelectedDispatch actions must have an event');
        }
        if (latestItems.current && latestItems.current.size > 0) {
          return instance.dispatch({
            type: 'select-none',
            event: action.event,
          });
        } else {
          if (fetchAllIds) {
            return instance.dispatch({
              type: 'select-all',
              ids: await fetchAllIds(),
              event: action.event,
            });
          }
          return instance.dispatch({ type: 'select-all', event: action.event });
        }
      }
      return instance.dispatch(action);
    },
    [instance.dispatch, fetchAllIds],
  );` 
</details>
<details>
  <summary>Code Path A</summary>
`  return instance.dispatch({
            type: 'select-none',
            event: action.event,
          });`
`
        case 'select-none':
          return { ...state, selectedItems: new Set<string>() };
`
Event is unused. 
</details>
<details>
  <summary>Code Path B</summary>
`  if (fetchAllIds) {
            return instance.dispatch({
              type: 'select-all',
              ids: await fetchAllIds(),
              event: action.event,
            });
          }
`
`
        case 'select-all':
          let selectedItems: string[] = [];
          if (action.ids && items && selectAllFilter) {
            const idsToInclude = new Set(
              items.filter(selectAllFilter).map(item => item.id),
            );
            selectedItems = action.ids.filter(id => idsToInclude.has(id));
          } else if (items && selectAllFilter) {
            selectedItems = items.filter(selectAllFilter).map(item => item.id);
          } else {
            selectedItems = action.ids || items.map(item => item.id);
          }
          return {
            ...state,
            selectedItems: new Set(selectedItems),
            selectedRange:
              action.ids && action.ids.length === 1
                ? { start: action.ids[0], end: null }
                : null,
          };
`
Event is unused. 
</details>
<details>
  <summary>Code Path C</summary>
` 
return instance.dispatch({ type: 'select-all', event: action.event })
`
Same Path as Path B Event is unused. 
</details>
<details>
  <summary>Code Path D</summary>
` return instance.dispatch(action);
`
`
       case 'select': {
`
We have Added a new check to ensure event is correctly set. 
&&
`
default:
          throw new Error('Unexpected action: ' + JSON.stringify(action));
`
Event is unused. 
</details>

Finally, the calls into useSelected 
<details>
<summary> Details </summary>
1.  PayeeTableHeader -> passes Event
2. PayeeTableRow -> passes Event
3. RuleRow -> passes Event
4. RulesHeader -> passes Event
5. DiscoverSchedulesTable -> passes Event
6. SimpleTransactionsTable -> passes Event && New code which doesnt
7. selectAscDesc -> passes Event
8. onUpdateAfterConfirm -> passes Event

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: Other option is we could create a "DummyEvent" but personal view is that is likely to cause bugs down the line.

Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: Ok, so I looked into this a bit more. It seems the underlying product problem we're trying to solve here is a lot more complicated actually. Apologies for continuing to send all these pieces of feedback, but I think that if we want to get this merged - it's important to get all the tiny details right so we wouldn't get bug reports about the feature.

So.. here's what I've noticed: the schedule creation modal does not respect the selected transactions.

https://github.com/actualbudget/actual/assets/886567/e230b938-236b-4d22-b26d-d28c90b3b930

We can think of the transaction table and schedule creation modal as two separate entities. The transaction table should pass the selected transaction ids to the creation modal. Next the user can interact with the schedule creation modal as usual: change some details AND select specific transactions in the list that is visible in the modal. We should preselect the transactions from the table. But the user can decide to un-select them there (which would only affect the modal, not the main table). And then they can click "create" which would link the new schedule with the transactions selected in the modal (not the table).

Lastly: we close the modal and "refresh" the state in the transaction table with the linked transactions. However, this should happen automatically without the need to dispatch any new event. You won't need any new logic for this.

One more thing to consider: if multiple transactions are selected - we cannot "blanket fill" the schedule creation modal. For multiple transactions it should use either the "one in" syntax or just leave the field blank (for example if different balances are selected).

Again: sorry for continuing to give feedback here. But I hope you see that the problem is a lot more complicated than we originally anticipated.

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: Not sure I fully get you, 
If more than one transaction is initially selected then it will use the first one in the "create schedule" helper, happy to add a tooltip highlighting this, but believe this makes sense as functionally, it doesn't make sense to create a single scheduled transaction from multiple transactions.
As you say, if we look to support creating a schedule from multiple transaction, it gets very muddy very quickly, and i don't think it would be very useful. 

Rest of it seems to work correctly from what i can see. 
https://github.com/actualbudget/actual/assets/30550215/7b84d381-34ae-4e23-8458-437dab9f9db6

Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: Ok, let me clarify in one sentence: Inside the schedule creation modal/page only the selected transactions should be linked to the schedule after it is creatd.

Currently this is not the case. That's a bug. All the other feedback I provided should help you solve this bug.

Screenshot 2024-01-21 at 17 00 33

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: I agree with your statement, and as per below video, only transactions selected on the "Schedule Creation Model" are linked to the newly created schedule. I can't see the bug you are detailing. 

See video:
https://github.com/actualbudget/actual/assets/30550215/d10b1eb5-69e8-457e-a3f6-722a98673ef7

Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx:279]: 1. one selected transaction, but later un-selected: yet it still gets attached to the schedule


https://github.com/actualbudget/actual/assets/886567/a33e6630-5111-44ce-b3cc-dc31b018502f


2. two selected transactions: only a single showing up in the modal


https://github.com/actualbudget/actual/assets/886567/5bac5258-d6ff-4e35-adc1-236b71fb3502


3. single transaction selected, but later the schedule config changed and multiple transactions selected: neither gets linked to the schedule


https://github.com/actualbudget/actual/assets/886567/081dff5e-984e-4819-ba60-d5b63b937ffe

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx]: ahah, you are correct, 
needed a fresh set of my eyes. 
Adjusted to fix and this cleans up the other comments you had as well. 
Let me know thoughts.

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx]: Adjusted it so the transaction you initially selected is always visible in the search trans box as well. My view, would be odd for you to select the tran to create from, but then not link it. 
Happy to remove if you disagree.

Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx]: Apologies for the slow reply; I've been on vacation.

-

I'm still able to reproduce most of [these bugs](https://github.com/actualbudget/actual/pull/2222#discussion_r1461031031). Your work here is definitely very much appreciated and I don't want to sound like I'm attacking your work, but I'd strongly recommend re-reading [this comment](https://github.com/actualbudget/actual/pull/2222#discussion_r1460976171). It encapsulates the idea how you could make the code cleaner while also fixing the bugs. Hope that helps!

Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx]: Hope you had a good vacation, 
No Problem on your response, Think I need to also explain myself better :). 

Going through the 
bugs you highlighted:
1) one selected transaction, but later un-selected: yet it still gets attached to the schedule - **Resolved**

https://github.com/actualbudget/actual/assets/30550215/7b02b25d-3b2d-46e0-809f-89f8ab81d40a

2) two selected transactions: only a single showing up in the modal - **Bypassed**
When selecting multiple transactions to create a new schedule from as you rightly highlighted it becomes messy & complex very quickly. Given in my view 95% of the time people are going to use this feature it will be by selecting a single transaction I have only implemented it to support this. As per comment here:

> If more than one transaction is initially selected then it will use the first one in the "create schedule" helper, happy to add a tooltip highlighting this, but believe this makes sense as functionally, it doesn't make sense to create a single scheduled transaction from multiple transactions.

I have updated this now, so the create option is only displayed when a single transaction is selected to simplify this, and avoid confusion from users. - See Video below **Happy to adjust this to disable the button instead with an explanation tooltip if you think that would be more suitable, I thought simply hiding the button looked cleaner**

https://github.com/actualbudget/actual/assets/30550215/6cea4ba8-44dc-4d3b-b9a5-f74009898ae0

3) This bug looks to be 2 different existing bugs you have triggered at the same time,  - **Existing Bug**

3a) When you change details in the add schedule model, previous selections are removed - See Video below highlighting it in the add schedule page. This should be fairly straight forward to fix but would make sense to do in another PR, happy to do this change. 

https://github.com/actualbudget/actual/assets/30550215/47cb17f2-ba5c-4bef-8b4b-0e650c155b45

3b) When a schedule is created and items are linked to it, if you toggle the select all on the transaction, the transaction seems to go into a "bad state" and even though the schedule shows the transactions as linked, the transactions themselves don't show as linked. 
Again, this looks to be an existing bug, that I am more than happy to tackle as a new PR. 
See video

https://github.com/actualbudget/actual/assets/30550215/ec58da0f-5404-4a4f-8530-cbc0dcbfff68

File: packages/desktop-client/src/hooks/useSelected.tsx (hunk 6 above)
Comment (xentara1) [reply packages/desktop-client/src/hooks/useSelected.tsx]: No longer changed

File: packages/desktop-client/src/hooks/useSelected.tsx (hunk 5 above)
Comment (MatissJanis) [reply packages/desktop-client/src/hooks/useSelected.tsx]: 1 - patched
2 - patched
3 - patched - but I think you misunderstood the bug report there. The issue was that when multiple transactions are selected in the modal (via select-all button) - neither of them would get linked to the schedule after the schedule creation.

The existing behaviour of resetting the selected transactions (in the modal) after the search parameters change is working as-expected. No need to change that.

File: packages/desktop-client/src/components/schedules/ScheduleDetails.jsx [hunk 8]
```diff
@@ -216,6 +222,18 @@ export function ScheduleDetails({ modalProps, actions, id }) {
...
+        if (fromTrans) {
+          schedule._account = transaction.account;
+          schedule._amount = transaction.amount;
+          schedule._amountOp = 'is';
+          schedule._date = {
+            frequency: 'monthly',
+            start: transaction.date,
+            patterns: [],
+          };
+          schedule.name = payees[transaction.payee].name;
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleDetails.jsx:234]: **🔨 warning:** ‏what if the transaction doesn't have a payee?

File: packages/desktop-client/src/components/schedules/ScheduleDetails.jsx [hunk 9]
```diff
@@ -321,7 +340,9 @@ export function ScheduleDetails({ modalProps, actions, id }) {
     };
   }, [state.schedule, state.transactionsMode, state.fields]);
 
-  const selectedInst = useSelected('transactions', state.transactions, []);
+  const selectedInst = useSelected('transactions', state.transactions, [
+    transaction ? transaction.id : null,
+  ]);
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleDetails.jsx:345]: **💬 suggestion:** ‏I'm not entirely sure what problems a null value here might cause in other parts of the product. It would be safer to do this:

```suggestion
  const selectedInst = useSelected('transactions', state.transactions,  transaction ?  [transaction.id] : []);
```

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx [hunk 10]
```diff
@@ -14,14 +17,23 @@ import { type CommonModalProps } from '../Modals';
...
   modalProps,
   actions,
   transactionIds: ids,
+  getTransaction,
+  pushModal,
 }: {
   actions: BoundActions;
   modalProps?: CommonModalProps;
   transactionIds: string[];
+  getTransaction: (tranID: string) => TransactionEntity;
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleLink.tsx:35]: **🥜 nitpick:** ‏no need to truncate the variable name
```suggestion
  getTransaction: (transactionId: string) => TransactionEntity;
```

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx [hunk 11]
```diff
@@ -70,6 +90,16 @@ export function ScheduleLink({
           value={filter}
           onChange={setFilter}
         />
+        {ids.length === 1 && (
+          <Button
+            type="primary"
+            style={{ marginLeft: 15, padding: '4px 10px' }}
+            onClick={onCreate}
+          >
+            <SvgAdd style={{ width: '20px', padding: '3px' }} />
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/schedules/ScheduleLink.tsx:99]: **🥜 nitpick:** ‏
```suggestion
            <SvgAdd style={{ width: 20, padding: 3 }} />
```

**Review by MatissJanis — CHANGES_REQUESTED** [review]
We're getting very close! Noticed a few other small bugs and some small code issues. LMK what you think.

File: packages/desktop-client/src/components/schedules/ScheduleDetails.jsx [hunk 12]
```diff
@@ -216,6 +222,18 @@ export function ScheduleDetails({ modalProps, actions, id }) {
...
+          schedule._amount = transaction.amount;
+          schedule._amountOp = 'is';
+          schedule._date = {
+            frequency: 'monthly',
+            start: transaction.date,
+            patterns: [],
+          };
+          schedule.name = payees[transaction.payee].name;
+          schedule._payee = transaction.payee;
+        }
```
[the remark this answers is not in the document]
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleDetails.jsx:236]: Done

File: packages/desktop-client/src/components/schedules/ScheduleDetails.jsx (hunk 8 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleDetails.jsx:234]: Done

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx (hunk 10 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleLink.tsx:35]: done

File: packages/desktop-client/src/components/schedules/ScheduleLink.tsx (hunk 11 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleLink.tsx:99]: done

File: packages/desktop-client/src/components/schedules/ScheduleDetails.jsx (hunk 9 above)
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleDetails.jsx:345]: Like it makes, sense

File: packages/desktop-client/src/components/schedules/ScheduleDetails.jsx [hunk 13]
```diff
@@ -143,6 +143,12 @@ export function ScheduleDetails({ modalProps, actions, id }) {
             fields: { ...state.fields, ...fields },
           };
         case 'set-transactions':
+          if (fromTrans) {
+            action.transactions = action.transactions.filter(
+              x => x.id !== transaction.id,
+            );
+            action.transactions.unshift(transaction);
+          }
```
[the remark this answers is not in the document]
Comment (xentara1) [reply packages/desktop-client/src/components/schedules/ScheduleDetails.jsx]: Done, However as discussed this now means that if you adjust the details of the schedule, you can lose the transaction you are trying to link from.

**Review by MatissJanis — APPROVED** [review]
LGTM! Thanks for being so patient and thanks for your contribution!
